src/SqlSquare/Parser.purs

```diff
@@ -0,0 +1,688 @@
+module SqlSquare.Parser
+  ( parse
+  , module SqlSquare.Parser.Tokenizer
+  ) where
+
+import Prelude
+
+import Control.Alt ((<|>))
+import Control.Monad.State (gets, put)
+import Control.MonadPlus (guard)
+
+import Data.Array as A
+import Data.NonEmpty ((:|))
+import Data.Either as E
+import Data.List ((:))
+import Data.List as L
+import Data.Foldable as F
+import Data.Maybe (Maybe(..), isNothing, fromMaybe, isJust, maybe)
+import Data.Json.Extended as EJ
+import Data.Path.Pathy as Pt
+import Data.String as S
+
+import SqlSquare.Utils ((∘), type (×), (×))
+import SqlSquare.Parser.Tokenizer (Token(..), tokenize, Literal(..))
+import SqlSquare.Signature as Sig
+import Matryoshka (class Corecursive, embed)
+
+import Text.Parsing.Parser as P
+import Text.Parsing.Parser.Combinators as PC
+import Text.Parsing.Parser.Pos (initialPos)
+
+parse ∷ ∀ t. Corecursive t (Sig.SqlF EJ.EJsonF) ⇒ String → E.Either P.ParseError t
+parse =
+  tokenize
+  >=> flip P.runParser do
+    res ← expr
+    eof
+    pure res
+
+token ∷ ∀ m. Monad m ⇒ P.ParserT (Array Token) m Token
+token = do
+  input ← gets \(P.ParseState input _ _) → input
+  case A.uncons input of
+    Nothing → P.fail "Unexpected EOF"
+    Just {head, tail} → do
+      put $ P.ParseState tail initialPos true
+      pure head
+
+eof ∷ ∀ m. Monad m ⇒ P.ParserT (Array Token) m Unit
+eof = do
+  input ← gets \(P.ParseState input _ _) → input
+  unless (A.null input) $ P.fail "Expected EOF"
+  pure unit
+
+whenTok ∷ ∀ m. Monad m ⇒ (Token → Boolean) → P.ParserT (Array Token) m Token
+whenTok f = PC.try do
+  a ← token
+  guard $ f a
+  pure a
+
+prod
+  ∷ ∀ a b m i
+  . Monad m
+  ⇒ P.ParserT i m a
+  → P.ParserT i m b
+  → (a → a → b → a)
+  → P.ParserT i m a
+prod pa pb f = do
+  l ← pa
+  p ← A.many do
+    o ← pb
+    r ← pa
+    pure $ o × r
+  pure $ F.foldl (\acc (o × r) → f acc r o) l p
+
+ident ∷ ∀ m. Monad m ⇒ P.ParserT (Array Token) m String
+ident = PC.try do
+  a ← token
+  case a of
+    Identifier s → pure s
+    _ → P.fail "Token is not an identifier"

```

Comment:
If you put `try` at the root of your parsers, you have no option to use it without `try`. I don't know if this flexibility is important, which is why I posed it as a question. I haven't written enough complex parsers to have a strong opinion. But it seems like you mostly want it in alternatives, and being able to tell in the alternative structure which is backtracking or not is helpful.